Python/chapter03/p06_animal_shelter/miguelHx.py

```diff
@@ -0,0 +1,222 @@
+        a.enqueue(d1, d2, d3, d4)
+        self.assertEqual(len(a), 4)
+        self.assertEqual(len(a.dogs), 4)
+        self.assertEqual(len(a.cats), 0)
+        self.assertEqual(d1.timestamp, 0)
+        self.assertEqual(d2.timestamp, 1)
+        self.assertEqual(d3.timestamp, 2)
+        self.assertEqual(d4.timestamp, 3)
+        # now, add a few cats...
+        c1 = Animal('Curio', 'cat')
+        c2 = Animal('Chai', 'cat')
+        c3 = Animal('Alma', 'cat')
+        c4 = Animal('Blanco', 'cat')
+        a.enqueue(c1, c2, c3, c4)
+        self.assertEqual(len(a), 8)
+        self.assertEqual(len(a.dogs), 4)
+        self.assertEqual(len(a.cats), 4)
+        self.assertEqual(c1.timestamp, 4)
+        self.assertEqual(c2.timestamp, 5)
+        self.assertEqual(c3.timestamp, 6)
+        self.assertEqual(c4.timestamp, 7)
+
+    def test_dequeue_dog_and_cat_queue(self):
+        a = AnimalShelter()
+        self.assertRaises(IndexError, lambda: a.dequeueDog())
+        self.assertRaises(IndexError, lambda: a.dequeueCat())
+        d1 = Animal('Lucky', 'dog')
+        d2 = Animal('Nova', 'dog')
+        d3 = Animal('Geralt', 'dog')
+        a.enqueue(d1, d2, d3)
+        self.assertEqual(len(a), 3)
+        self.assertEqual(len(a.dogs), 3)
+        self.assertEqual(len(a.cats), 0)
+        d = a.dequeueDog()
+        self.assertEqual(len(a), 2)
+        self.assertEqual(len(a.dogs), 2)
+        self.assertEqual(len(a.cats), 0)
+        self.assertEqual(d1, d)
+        d = a.dequeueDog()
+        self.assertEqual(len(a), 1)
+        self.assertEqual(len(a.dogs), 1)
+        self.assertEqual(len(a.cats), 0)
+        self.assertEqual(d2, d)
+        d = a.dequeueDog()
+        self.assertEqual(len(a), 0)
+        self.assertEqual(len(a.dogs), 0)
+        self.assertEqual(len(a.cats), 0)
+        self.assertEqual(d3, d)
+        self.assertRaises(IndexError, lambda: a.dequeueDog())
+
+    def test_dequeue_any(self):
+        a = AnimalShelter()
+        self.assertRaises(IndexError, lambda: a.dequeueAny())
+        # scenario 1: Only dogs
+        d1 = Animal('Lucky', 'dog')
+        d2 = Animal('Nova', 'dog')
+        d3 = Animal('Geralt', 'dog')
+        a.enqueue(d1, d2, d3)
+        self.assertEqual(len(a), 3)
+        self.assertEqual(len(a.dogs), 3)
+        self.assertEqual(len(a.cats), 0)
+        d = a.dequeueAny()
+        self.assertEqual(len(a), 2)
+        self.assertEqual(len(a.dogs), 2)
+        self.assertEqual(len(a.cats), 0)
+        self.assertEqual(d1, d)
+        d = a.dequeueAny()
+        self.assertEqual(len(a), 1)
+        self.assertEqual(len(a.dogs), 1)
+        self.assertEqual(len(a.cats), 0)
+        self.assertEqual(d2, d)
+        d = a.dequeueAny()
+        self.assertEqual(len(a), 0)
+        self.assertEqual(len(a.dogs), 0)
+        self.assertEqual(len(a.cats), 0)
+        self.assertEqual(d3, d)
+        self.assertRaises(IndexError, lambda: a.dequeueAny())
+        # scenario 2: Only cats
```

Comment:
given that every scenario is (logically) independent of the previous ones, it may be better to split this test in four for readability. also, if the test fails, a maintainer would need to look at the test code and read the position of the failure, plus fix the failure to discover whether the following behaviors exercised in the test still work.

It is a best practice to make tests as small and isolated as possible. Per https://dzone.com/articles/unit-testing-best-practices-how-to-get-the-most-ou

> #### Unit Tests Should Verify a Single-Use Case
> Good tests validate one thing and one thing only, which means that typically, they validate a single use-case. Tests that follow this best practice are simpler and more understandable, and that is good for maintainability and debugging. Tests that validate more than one thing can easily become complex and time-consuming to maintain. Don't let this happen.

Reply:
Great feedback, thank you!  Updated here:  a88c163df248d3ac2735534ba40ef02e3af68626